File: src/jump_game/jump_game.cc
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    bool canJump(vector<int> &nums);
};


bool Solution::canJump(vector<int> &nums) {
    int reach = 0, len = nums.size();
    for (int i = 0; i < len && i <= reach; i++) {
        reach = max(i + nums[i], reach);
    }
    return reach > len;
}

int main(int argc, char **argv) {

    Solution s;

    vector<int> v1{2, 3, 1, 1, 4};
    cout << "v1: " << s.canJump(v1);


    return 0;
}

File: src/jump_game/jump_game.rs
fn main() {
    let v = vec![2, 3, 1, 1, 4];

    println!("can jump: {}", Solution::can_jump(v));
}



struct Solution;

impl Solution {
    pub fn can_jump(nums: Vec<usize>) -> bool {
        let mut reach: usize = 0;
        let len: usize = nums.len();
        let mut index: usize = 0;
        while index < len && index <= reach {
            reach = if (index + nums[index]) > reach {
                index + nums[index]
            } else {
                reach
            };
            index += 1;
        }
        reach > len
    }
}